implementations/rust/ockam/ockam_node/src/storage/database/sqlx_types.rs

```diff
@@ -0,0 +1,200 @@
+/// to an Sqlite database. There is a 1 to 1 mapping with the database native types
+impl Encode<'_, Sqlite> for SqlxType {
+    fn encode_by_ref(&self, buf: &mut <Sqlite as HasArguments>::ArgumentBuffer) -> IsNull {
+        match self {
+            SqlxType::Text(v) => <String as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+            SqlxType::Blob(v) => <Vec<u8> as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+            SqlxType::Integer(v) => <i64 as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+            SqlxType::Real(v) => <f64 as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+        }
+    }
+
+    fn produces(&self) -> Option<<Sqlite as Database>::TypeInfo> {
+        Some(match self {
+            SqlxType::Text(_) => <String as Type<Sqlite>>::type_info(),
+            SqlxType::Blob(_) => <Vec<u8> as Type<Sqlite>>::type_info(),
+            SqlxType::Integer(_) => <i64 as Type<Sqlite>>::type_info(),
+            SqlxType::Real(_) => <f64 as Type<Sqlite>>::type_info(),
+        })
+    }
+}
+
+/// This trait can be implemented by any type that can be converted to a database type
+/// Typically an `Identifier` (to a `Text`), a `TimestampInSeconds` (to an `Integer`) etc...
+///
+/// This allows a value to be used as a bind parameters in a sqlx query for example:
+///
+/// use std::str::FromStr;
+/// use sqlx::query_as;
+/// use ockam_node::database::{SqlxType, ToSqlxType};
+///
+/// // newtype for a UNIX-like timestamp
+/// struct TimestampInSeconds(u64);
+///
+/// // this implementation maps the TimestampInSecond type to one of the types that Sqlx
+/// // can serialize for sqlite
+/// impl ToSqlxType for TimestampInSeconds {
+///     fn to_sql(&self) -> SqlxType {
+///         self.0.to_sql()
+///     }
+/// }
+///
+/// let timestamp = TimestampInSeconds(10000000);
+/// let query = query_as("SELECT * FROM identity WHERE created_at >= $1").bind(timestamp.as_sql());
+///
+///
+pub trait ToSqlxType {
+    /// Return the appropriate sql type
+    fn to_sql(&self) -> SqlxType;
+}
+
+impl ToSqlxType for String {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Text(self.clone())
+    }
+}
+
+impl ToSqlxType for &str {
+    fn to_sql(&self) -> SqlxType {
+        self.to_string().to_sql()
+    }
+}
+
+impl ToSqlxType for bool {
+    fn to_sql(&self) -> SqlxType {
+        if *self {
+            1.to_sql()
+        } else {
+            0.to_sql()
+        }
+    }
+}
+
+impl ToSqlxType for u64 {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Integer(*self as i64)
+    }
+}
+
+impl ToSqlxType for u32 {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Integer(*self as i64)
+    }
+}
+
+impl ToSqlxType for u16 {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Integer(*self as i64)
+    }
+}
+
+impl ToSqlxType for u8 {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Integer(*self as i64)
+    }
+}
+
+impl ToSqlxType for usize {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Integer(*self as i64)
```

Comment:
That's also potentially unsafe

Reply:
I suggest we remove it. I don't see a reasoning of why we would store a type of an architecture-dependent size in the db. Let's force anyone who is trying to save it to do a proper conversion to a fixed size integer (and make sure it's correct).

Reply:
Done. There was an `id` field in `ProjectUserRole` using `usize` but I replaced it with `u64`.